test/show.jl

```diff
@@ -65,151 +64,116 @@ end
-    ├─────┼──────────┼──────────┼──────────┤
-    │ 1   │ 10000001 │ 10000026 │ 10000051 │
-    │ 2   │ 10000002 │ 10000027 │ 10000052 │
-    │ 3   │ 10000003 │ 10000028 │ 10000053 │
-    │ 4   │ 10000004 │ 10000029 │ 10000054 │
-    │ 5   │ 10000005 │ 10000030 │ 10000055 │
-    │ 6   │ 10000006 │ 10000031 │ 10000056 │
-    │ 7   │ 10000007 │ 10000032 │ 10000057 │
-    │ 8   │ 10000008 │ 10000033 │ 10000058 │
-    │ 9   │ 10000009 │ 10000034 │ 10000059 │
-    │ 10  │ 10000010 │ 10000035 │ 10000060 │
-    │ 11  │ 10000011 │ 10000036 │ 10000061 │
-    │ 12  │ 10000012 │ 10000037 │ 10000062 │
-    │ 13  │ 10000013 │ 10000038 │ 10000063 │
-    │ 14  │ 10000014 │ 10000039 │ 10000064 │
-    │ 15  │ 10000015 │ 10000040 │ 10000065 │
-    │ 16  │ 10000016 │ 10000041 │ 10000066 │
-    │ 17  │ 10000017 │ 10000042 │ 10000067 │
-    │ 18  │ 10000018 │ 10000043 │ 10000068 │
-    │ 19  │ 10000019 │ 10000044 │ 10000069 │
-    │ 20  │ 10000020 │ 10000045 │ 10000070 │
-    │ 21  │ 10000021 │ 10000046 │ 10000071 │
-    │ 22  │ 10000022 │ 10000047 │ 10000072 │
-    │ 23  │ 10000023 │ 10000048 │ 10000073 │
-    │ 24  │ 10000024 │ 10000049 │ 10000074 │
-    │ 25  │ 10000025 │ 10000050 │ 10000075 │"""
+        25×5 DataFrame
+         Row │ x1        x2        x3        x ⋯
+             │ Int64     Int64     Int64     I ⋯
+        ─────┼──────────────────────────────────
+           1 │ 10000001  10000026  10000051  1 ⋯
+           2 │ 10000002  10000027  10000052  1
+           3 │ 10000003  10000028  10000053  1
+           4 │ 10000004  10000029  10000054  1
+           5 │ 10000005  10000030  10000055  1 ⋯
+           6 │ 10000006  10000031  10000056  1
+           7 │ 10000007  10000032  10000057  1
+           8 │ 10000008  10000033  10000058  1
+           9 │ 10000009  10000034  10000059  1 ⋯
+          10 │ 10000010  10000035  10000060  1
+          11 │ 10000011  10000036  10000061  1
+          12 │ 10000012  10000037  10000062  1
+          13 │ 10000013  10000038  10000063  1 ⋯
+          14 │ 10000014  10000039  10000064  1
+          15 │ 10000015  10000040  10000065  1
+          16 │ 10000016  10000041  10000066  1
+          17 │ 10000017  10000042  10000067  1 ⋯
+          18 │ 10000018  10000043  10000068  1
+          19 │ 10000019  10000044  10000069  1
+          20 │ 10000020  10000045  10000070  1
+          21 │ 10000021  10000046  10000071  1 ⋯
+          22 │ 10000022  10000047  10000072  1
+          23 │ 10000023  10000048  10000073  1
+          24 │ 10000024  10000049  10000074  1
+          25 │ 10000025  10000050  10000075  1 ⋯
+                               2 columns omitted"""
 end
 
 @testset "IOContext parameters test" begin
     df = DataFrame(A = Int64[1:4;], B = ["x\"", "∀ε>0: x+ε>x", "z\$", "A\nC"],
                    C = Float32[1.0, 2.0, 3.0, 4.0])
-    str1, size = capture_stdout() do
-        show(df)
+    str1 = capture_stdout() do
+        show(df, screen_size = (24,80))
     end
```

Comment:
> I think the point of this test was to check that we adapt to the REPL's size automatically. Why do you need to change it? Also why does PrettyTables take `screen_size` as a separate argument instead of using `:displaysize`?

This is a problem I found with Julia (or a misconception I have about IOs). When `io` is a file, then `displaysize(io)` returns the values of env. variables `LINES` and `COLUMNS`. If they are not defined, then it returns `(24,80)`. With a file, I do not want to output to be cropped. Hence, I use the following code to identify this:

```julia
withenv("LINES" => -1, "COLUMNS" => -1) do
    screen.size = displaysize(io)
end
```

Thus, if `io` is a file, then PrettyTables will not crop the output as expected.

```julia
julia> f = open("teste.txt", "w")
IOStream(<file teste.txt>)

julia> displaysize(f)
(24, 80)

julia> withenv("LINES" => -1, "COLUMNS" => -1) do
           println(displaysize(f))
       end
(-1, -1)
```

Notice that everything is fine for usual IOs that are not files:

```julia
julia> buf = IOBuffer()
IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1)

julia> io = IOContext(buf, :displaysize => (100,100))
IOContext(IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1))

julia> withenv("LINES" => -1, "COLUMNS" => -1) do
           println(displaysize(io))
       end
(100, 100)
```

The problem here is how `capture_output` changes the `stdout`. In this case, it is also returning the values of those environment variables:

```julia
julia> function capture_stdout(f::Function)
           oldstdout = stdout
           rd, wr = redirect_stdout()
           f()
           redirect_stdout(oldstdout)
           close(wr)
           str = read(rd, String)
           close(rd)
           str
       end
capture_stdout (generic function with 1 method)

julia> str1 = capture_stdout() do
           withenv("LINES" => -1, "COLUMNS" => -1) do
               println(displaysize(stdout))
           end
       end
"(-1, -1)\n"

julia> str1 = capture_stdout() do
           withenv("LINES" => 10, "COLUMNS" => 20) do
               println(displaysize(stdout))
           end
       end
"(10, 20)\n"
```

Thus, for PrettyTables.jl, we are passing a file and the output must never be cropped, unless I force cropping by manually setting the variable `screen_size`.

> Also why does PrettyTables take screen_size as a separate argument instead of using :displaysize?

This is just to force the output to be cropped. If `screen_size` is `nothing` (default), then it uses the output of `displaysize` considering the discussion about the env. variables `LINES` and `COLUMNS`.

Reply:
I thought it is safest to make decision based on:
>    •    :limit: Boolean specifying that containers should be truncated, e.g. showing … in place of most elements.
>    •    :displaysize: A Tuple{Int,Int} giving the size in rows and columns to use for text output. This can be
        used to override the display size for called functions, but to get the size of the screen use the
        displaysize function.

(taken from `IOContext` docstring)

Reply:
To develop on what @bkamins said, the solution we use in the current code is do check `get(io, :limit, false)` first. It will return `false` on files, so in that case we don't have to check the value of `displaysize`. I think this is the recommended way to handle this in Julia.

Reply:
I really tried to use this parameters from `IOContext`, but it does not work properly for `stdout`, which is the most common case:

```julia
julia> get(stdout, :limit, false)
false
```

In DataFrames.jl, IIRC, you change the default of the `get` to `true` if the user does not pass the `io`. Hence, I really think we cannot rely only in the `:limit` property of `IOContext`.

Reply:
IIRC the REPL doesn't use `stdout` directly, but it wraps it in an `IOContext` with `:limit => true`, right?

Reply:
I have no idea. Do you have any advice how should I change PrettyTables.jl to handle this?

Reply:
This is what I understand happens. @ronisbr - can you please check. This is a difference between:
```
x = rand(100,100)
```
(`x` is cropped in REPL)
and doing
```
show(x)
```
in REPL (which is not cropped)

Reply:
In PrettyTables.jl (or in DataFrames.jl wrapper - whichever is more convenient) a check for `:limit` property should be made - just as we currently do in legacy dipslay for data frames.

Reply:
Hum, very nice this example using `show`. I will do the following: if the user does not pass the `io`, then I will wrap `stdout` in an `IOContext` with the `:limit` set to `true`. Otherwise, I will always use `:limit` and `:displaysize`. However, I will make this with a priority lower than the options `crop` and `screen_size`. Thus, the user can easily change the behavior during the call of `pretty_tables`.

Reply:
Actually not, I think `:limit` must have a higher priority. If `:limit` is `false`, then I cannot trust in `:displaysize`, then I will only allow cropping if the user sets `screen_size`. The algorithm will be:

1. If `:limit` is `true` and `:displaysize` is positive , then crop the output if `crop != :none`.
2. If `:limit` is `false` and `screen_size` is not specified, then do not crop the output.
3. If `screen_size` is specified, then always crop the output if `crop != :none`.

Reply:
Handling each pair of properties separately might be the simplest solution:
- have `crop = get(io, :limit, false) ? :both : :none`
- have `screen_size` default to `displaysize(io)`, and rename it to `displaysize`

And then just use the arguments as currently. If the user sets `crop != :none` then it's their responsibility to ensure `displaysize` has a correct value (either via the `IOContext` or via `displaysize`), just like currently.

Reply:
Done! With the modifications to PrettyTables.jl, now I can remove the `screen_size = (24,80)`.